### PR TITLE
[SPARK-25313][SQL]Fix regression in FileFormatWriter output names

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

Let's see the follow example:
```
        val location = "/tmp/t"
        val df = spark.range(10).toDF("id")
        df.write.format("parquet").saveAsTable("tbl")
        spark.sql("CREATE VIEW view1 AS SELECT id FROM tbl")
        spark.sql(s"CREATE TABLE tbl2(ID long) USING parquet location $location")
        spark.sql("INSERT OVERWRITE TABLE tbl2 SELECT ID FROM view1")
        println(spark.read.parquet(location).schema)
        spark.table("tbl2").show()
```
The output column name in schema will be `id` instead of `ID`, thus the last query shows nothing from `tbl2`. 
By enabling the debug message we can see that the output naming is changed from `ID` to `id`, and then the `outputColumns` in `InsertIntoHadoopFsRelationCommand` is changed in `RemoveRedundantAliases`.
wechatimg5

wechatimg4

**To guarantee correctness**, we should change the output columns from `Seq[Attribute]` to `Seq[String]` to avoid its names being replaced by optimizer.

I will fix project elimination related rules in https://github.com/apache/spark/pull/22311 after this one.

## How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by JoshRosen** [comment]
(This is a test comment to test a GitHub Integration; please ignore)

**Comment by gengliangwang** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala [hunk 1]
```diff
@@ -38,6 +38,20 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
...
+  def outputWithNames(names: Seq[String]): Seq[Attribute] = {
+    // Save the output attributes to a variable to avoid duplicated function calls.
+    val outputAttributes = output
+    assert(outputAttributes.length == names.length,
+      "The length of provided names doesn't match the length of output attributes.")
+    outputAttributes.zipWithIndex.map { case (element, index) =>
+      element.withName(names(index))
+    }
+  }
+
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: If #22311 merged, we don't need this function anymore? If so, IMHO it'd be better to fix this issue in the `FileFormatWriter` side as a workaround?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: +1

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: or make it a util function

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 2]
```diff
@@ -495,7 +496,9 @@ case class DataSource(
               s"Unable to resolve $name given [${data.output.map(_.name).mkString(", ")}]")
           }
         }
-        val resolved = cmd.copy(partitionColumns = resolvedPartCols, outputColumns = outputColumns)
+        val resolved = cmd.copy(
+          partitionColumns = resolvedPartCols,
+          outputColumnNames = outputColumns.map(_.name))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: why can't we use `outputColumnNames` directly here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: It seems overkill to add a function here. But in `FileFormatWriter` we can't not access `LogicalPlan` to get the attributes.
Another way is to put this method in a Util.
Do you have a good suggestion?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: I was thinking...
```
object FileFormatWriter {
  ...

  // workaround: a helper function...
  def outputWithNames(outputAttributes: Seq[Attribute], names: Seq[String]): Seq[Attribute] = {
     assert(outputAttributes.length == names.length,
       "The length of provided names doesn't match the length of output attributes.")
     outputAttributes.zipWithIndex.map { case (element, index) =>
       element.withName(names(index))
     }
   }
```
Then, in each callsite, just say `FileFormatWriter. outputWithNames(logicalPlan.output, names)`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/DataWritingCommand.scala [hunk 3]
```diff
@@ -53,3 +57,21 @@ trait DataWritingCommand extends Command {
...
+object DataWritingCommand {
+  /**
+   * Returns output attributes with provided names.
+   * The length of provided names should be the same of the length of [[LogicalPlan.output]].
+   */
+  def logicalPlanOutputWithNames(
+      query: LogicalPlan,
+      names: Seq[String]): Seq[Attribute] = {
+    // Save the output attributes to a variable to avoid duplicated function calls.
+    val outputAttributes = query.output
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/DataWritingCommand.scala]: `query: LogicalPlan` -> `outputAttributes: Seq[Attribute]` in the function argument, then drop the line above?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -2853,6 +2854,81 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+        val expectedSchema = StructType(Seq(
+          StructField("COL1", LongType, true),
+          StructField("COL3", IntegerType, true),
+          StructField("COL2", IntegerType, true)))
+        assert(spark.read.parquet(location).schema == expectedSchema)
+        checkAnswer(spark.table("tbl2"), df)
+      }
+    }
+  }
+
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: better to move these tests into `DataFrameReaderWriterSuite`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 5]
```diff
@@ -69,7 +69,7 @@ case class InsertIntoHiveTable(
     query: LogicalPlan,
     overwrite: Boolean,
     ifPartitionNotExists: Boolean,
-    outputColumns: Seq[Attribute]) extends SaveAsHiveFile {
+    outputColumnNames: Seq[String]) extends SaveAsHiveFile {
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: For better test coverage, can you add tests for hive tables?

Comment (gengliangwang) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: No problem 👍

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/DataWritingCommand.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/DataWritingCommand.scala]: I think both are OK. The current way makes it easier to call this Util function, while the ways you suggests makes the argument carrying minimal information.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 6]
```diff
@@ -24,6 +24,7 @@ import java.util.concurrent.atomic.AtomicBoolean
 
 import org.apache.spark.{AccumulatorSuite, SparkException}
 import org.apache.spark.scheduler.{SparkListener, SparkListenerJobStart}
+import org.apache.spark.sql.catalyst.TableIdentifier
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: unnecessary change

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 7]
```diff
@@ -805,6 +805,81 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
...
+  test("Insert overwrite table command should output correct schema: complex") {
+    withTable("tbl", "tbl2") {
+      withView("view1") {
+        val df = spark.range(10).map(x => (x, x.toInt, x.toInt)).toDF("col1", "col2", "col3")
+        df.write.format("parquet").saveAsTable("tbl")
+        spark.sql("CREATE VIEW view1 AS SELECT * FROM tbl")
+        spark.sql("CREATE TABLE tbl2(COL1 long, COL2 int, COL3 int) USING parquet PARTITIONED " +
+          "BY (COL2) CLUSTERED BY (COL3) INTO 3 BUCKETS")
+        spark.sql("INSERT OVERWRITE TABLE tbl2 SELECT COL1, COL2, COL3 " +
+          "FROM view1 CLUSTER BY COL3")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: is it legal to put `CLUSTER BY` in the INSERT statement?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 8]
```diff
@@ -754,6 +754,47 @@ class HiveDDLSuite
     }
   }
 
+  test("Insert overwrite Hive table should output correct schema") {
+    withTable("tbl", "tbl2") {
+      withView("view1") {
+        spark.sql("CREATE TABLE tbl(id long)")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: please run this test within `withSQLConf(HiveUtils.CONVERT_METASTORE_PARQUET -> false)`

**Comment by cloud-fan** [comment]
LGTM except some minor comments

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 8 above)
Comment (gengliangwang) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: I am not familiar with Hive. But as I look at the debug message of this logical plan, the top level is `InsertIntoHiveTable `default`.`tbl2`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, true, false, [ID]`. It should not be related to this configuration, right?

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 9]
```diff
@@ -805,6 +805,80 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
...
 
+  test("Insert overwrite table command should output correct schema: basic") {
+    withTable("tbl", "tbl2") {
+      withView("view1") {
+        val df = spark.range(10).toDF("id")
+        df.write.format("parquet").saveAsTable("tbl")
+        spark.sql("CREATE VIEW view1 AS SELECT id FROM tbl")
+        spark.sql("CREATE TABLE tbl2(ID long) USING parquet")
+        spark.sql("INSERT OVERWRITE TABLE tbl2 SELECT ID FROM view1")
+        val identifier = TableIdentifier("tbl2", Some("default"))
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: `default` is the default database name, isn't it? I'd remove it from the test or use `spark.catalog.currentDatabase`.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 10]
```diff
@@ -805,6 +805,80 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
...
+  test("Insert overwrite table command should output correct schema: complex") {
+    withTable("tbl", "tbl2") {
+      withView("view1") {
+        val df = spark.range(10).map(x => (x, x.toInt, x.toInt)).toDF("col1", "col2", "col3")
+        df.write.format("parquet").saveAsTable("tbl")
+        spark.sql("CREATE VIEW view1 AS SELECT * FROM tbl")
+        spark.sql("CREATE TABLE tbl2(COL1 long, COL2 int, COL3 int) USING parquet PARTITIONED " +
+          "BY (COL2) CLUSTERED BY (COL3) INTO 3 BUCKETS")
+        spark.sql("INSERT OVERWRITE TABLE tbl2 SELECT COL1, COL2, COL3 FROM view1")
+        val identifier = TableIdentifier("tbl2", Some("default"))
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Same as above.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 11]
```diff
@@ -805,6 +805,80 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
...
+        val location = spark.sessionState.catalog.getTableMetadata(identifier).location.toString
+        val expectedSchema = StructType(Seq(
+          StructField("COL1", LongType, true),
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: `nullable` is `true` by default.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 12]
```diff
@@ -805,6 +805,80 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
...
+          StructField("COL3", IntegerType, true),
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: You could use a little magic here: `$"COL1".int`

```
scala> $"COL1".int
res1: org.apache.spark.sql.types.StructField = StructField(COL1,IntegerType,true)
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala [hunk 13]
```diff
@@ -63,7 +63,7 @@ case class CreateHiveTableAsSelectCommand(
         query,
         overwrite = false,
         ifPartitionNotExists = false,
-        outputColumns = outputColumns).run(sparkSession, child)
+        outputColumnNames = outputColumnNames).run(sparkSession, child)
```
Comment (jaceklaskowski) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala]: Can you remove one `outputColumnNames`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 14]
```diff
@@ -754,6 +754,47 @@ class HiveDDLSuite
...
+      }
+    }
+  }
+
+  test("Insert into Hive directory should output correct schema") {
+    withTable("tbl") {
+      withView("view1") {
+        withTempPath { path =>
+          spark.sql("CREATE TABLE tbl(id long)")
+          spark.sql("INSERT OVERWRITE TABLE tbl SELECT 4")
```
Comment (jaceklaskowski) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: `s/SELECT/VALUES` as it could be a bit more Spark-idiomatic?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala (hunk 5 above)
Comment (maropu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 15]
```diff
@@ -754,6 +754,47 @@ class HiveDDLSuite
...
 
+  test("Insert overwrite Hive table should output correct schema") {
+    withTable("tbl", "tbl2") {
+      withView("view1") {
+        spark.sql("CREATE TABLE tbl(id long)")
+        spark.sql("INSERT OVERWRITE TABLE tbl SELECT 4")
+        spark.sql("CREATE VIEW view1 AS SELECT id FROM tbl")
+        spark.sql("CREATE TABLE tbl2(ID long)")
+        spark.sql("INSERT OVERWRITE TABLE tbl2 SELECT ID FROM view1")
+        checkAnswer(spark.table("tbl2"), Seq(Row(4)))
```
Comment (wangyum) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Add schema assert please. We can read data since [SPARK-25132](https://issues.apache.org/jira/browse/SPARK-25132).

**Comment by maropu** [comment]
retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 15 above)
Comment (gengliangwang) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Good point. I found that `CreateHiveTableAsSelectCommand` output wrong schema after adding a new test case.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala [hunk 16]
```diff
@@ -63,13 +63,14 @@ case class CreateHiveTableAsSelectCommand(
...
     } else {
       // TODO ideally, we should get the output data ready first and then
       // add the relation into catalog, just in case of failure occurs while data
       // processing.
       assert(tableDesc.schema.isEmpty)
-      catalog.createTable(tableDesc.copy(schema = query.schema), ignoreIfExists = false)
+      val schema = DataWritingCommand.logicalPlanSchemaWithNames(query, outputColumnNames)
+      catalog.createTable(tableDesc.copy(schema = schema), ignoreIfExists = false)
```
Comment (gengliangwang) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala]: The schema naming need to be consistent with `outputColumnNames` here.

**Review by jaceklaskowski — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 17]
```diff
@@ -56,7 +56,7 @@ case class InsertIntoHadoopFsRelationCommand(
     mode: SaveMode,
     catalogTable: Option[CatalogTable],
     fileIndex: Option[FileIndex],
-    outputColumns: Seq[Attribute])
+    outputColumnNames: Seq[String])
   extends DataWritingCommand {
   import org.apache.spark.sql.catalyst.catalog.ExternalCatalogUtils.escapePathName
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: Line 66: `query.schema` should be `DataWritingCommand.logicalPlanSchemaWithNames(query, outputColumnNames)`.

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: Oh, then we can use this method instead.
```
def checkColumnNameDuplication(
      columnNames: Seq[String], colType: String, caseSensitiveAnalysis: Boolean): Unit
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 18]
```diff
@@ -805,6 +805,80 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
     }
   }
 
+  test("Insert overwrite table command should output correct schema: basic") {
+    withTable("tbl", "tbl2") {
+      withView("view1") {
+        val df = spark.range(10).toDF("id")
```
[the remark this answers is not in the document]
Comment (jaceklaskowski) [reply sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: "case sensitive"? How is so since Spark SQL is case-insensitive by default?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala [hunk 19]
```diff
@@ -82,7 +83,7 @@ case class CreateHiveTableAsSelectCommand(
           query,
           overwrite = true,
           ifPartitionNotExists = false,
-          outputColumns = outputColumns).run(sparkSession, child)
+          outputColumnNames = outputColumnNames).run(sparkSession, child)
```
Comment (jaceklaskowski) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala]: Why is this duplication needed here?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 20]
```diff
@@ -754,6 +754,54 @@ class HiveDDLSuite
     }
   }
 
+  test("Insert overwrite Hive table should output correct schema") {
+    withSQLConf(CONVERT_METASTORE_PARQUET.key -> "false") {
+      withTable("tbl", "tbl2") {
+        withView("view1") {
+          spark.sql("CREATE TABLE tbl(id long)")
+          spark.sql("INSERT OVERWRITE TABLE tbl VALUES 4")
```
Comment (jaceklaskowski) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: I might be missing something, but why does this test use SQL statements not DataFrameWriter API, e.g. `Seq(4).toDF("id").write.mode(SaveMode.Overwrite).saveAsTable("tbl")`?

**Review by jaceklaskowski — CHANGES_REQUESTED** [review]

**Comment by wangyum** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala (hunk 19 above)
Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala]: what's the duplication?

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala]: I feel it's better to specify parameters by name if the previous parameter is already specified by name, e.g. `ifPartitionNotExists = false`

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by wangyum** [comment]
@gengliangwang We need backport this pr to branch-2.3.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala [hunk 21]
```diff
@@ -38,6 +38,20 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
...
+   * Returns output attributes with provided names.
+   * The length of provided names should be the same of the length of [[output]].
+   */
+  def outputWithNames(names: Seq[String]): Seq[Attribute] = {
+    // Save the output attributes to a variable to avoid duplicated function calls.
+    val outputAttributes = output
+    assert(outputAttributes.length == names.length,
+      "The length of provided names doesn't match the length of output attributes.")
+    outputAttributes.zipWithIndex.map { case (element, index) =>
+      element.withName(names(index))
```
[the remark this answers is not in the document]
Comment (xyiswinner) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: In what situations would `outputAttributes.length != names.length` happen
